File: src/c/quickselect.c
/**
 * @file    algoB_C_quickselect.c
 * @details LeetCode 215 · Kth Largest Element in an Array — Quickselect (3-way partition)
 * @algorithm Treat the k-th largest element as the element at index (k-1)
 *            in a descending order. Repeatedly choose a random pivot and
 *            partition the current range into three parts: > pivot, == pivot,
 *            and < pivot. Then narrow the search interval to the part that
 *            can still contain the k-th largest position. Since we never
 *            fully sort the array, the expected running time is O(N).
 * @author  Team 5 이영민, 임재윤
 * @date    2025-11-22
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Swap two integers by pointer
void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

/**
 * 3-way partition around a randomly chosen pivot.
 *
 * After partitioning the subarray nums[left..right], we have:
 *   nums[left .. lt-1]  : elements strictly greater than pivot
 *   nums[lt   .. gt]    : elements equal to pivot
 *   nums[gt+1 .. right] : elements strictly less than pivot
 *
 * The indices lt and gt are returned by reference.
 */
void partition3(int* nums, int left, int right, int* lt, int* gt) {
    // Choose a random pivot index in [left, right]
    int pivotIndex = left + rand() % (right - left + 1);
    int pivot = nums[pivotIndex];

    // Move pivot to the front for convenience
    swap(&nums[left], &nums[pivotIndex]);

    int i = left;
    *lt = left;      // next position to place an element > pivot
    *gt = right;     // next position to place an element < pivot

    // Dutch National Flag partitioning
    while (i <= *gt) {
        if (nums[i] > pivot) {
            // For k-th "largest", values greater than pivot go to the left side
            swap(&nums[i], &nums[*lt]);
            (*lt)++;
            i++;
        } else if (nums[i] < pivot) {
            // Values smaller than pivot go to the right side
            swap(&nums[i], &nums[*gt]);
            (*gt)--;
            // do not increment i here, because the swapped-in element at i
            // still needs to be examined
        } else {
            // nums[i] == pivot → stay in the middle region
            i++;
        }
    }
}

/**
 * Quickselect to find the k-th largest element (1-based k) in-place.
 *
 * nums  : array of integers
 * left  : left boundary of the current search range
 * right : right boundary of the current search range
 * k     : rank of the desired element in descending order (1-based)
 */
int quickSelect(int* nums, int left, int right, int k) {
    // Target index in "descending" order is (k - 1) in 0-based indexing
    int targetIndex = k - 1;

    while (left <= right) {
        int lt, gt;
        partition3(nums, left, right, &lt, &gt);

        // Region mapping (in descending order):
        //   [left .. lt-1]  : indices 0 .. (lt-1-left)
        //   [lt   .. gt]    : all equal to pivot
        //   [gt+1 .. right] : smaller than pivot

        if (targetIndex < lt) {
            // k-th largest lies in the "greater than pivot" region
            right = lt - 1;
        } else if (targetIndex > gt) {
            // k-th largest lies in the "less than pivot" region
            left = gt + 1;
        } else {
            // targetIndex is within [lt .. gt] (== pivot region)
            // any element in this region has the correct rank
            return nums[lt];
        }
    }
    // Should not be reached if k is valid
    return -1;
}

int findKthLargest(int* nums, int numsSize, int k) {
    // Initialize random seed so that pivot selection varies between runs
    srand((unsigned int)time(NULL));

    // Run Quickselect on the full range
    return quickSelect(nums, 0, numsSize - 1, k);
}

File: src/python/quickselect.py
"""
@file     algoB_Python_quickselect.py
@details  LeetCode 215 · Kth Largest Element in an Array — Quickselect (3-way partition)
@algorithm View the k-th largest element as the element at index (len(nums) - k)
           in ascending order. Repeatedly pick a random pivot, partition the
           current range into three regions (< pivot, == pivot, > pivot), and
           continue the search only in the region that contains this target
           index. Since we never fully sort the array, the expected running
           time is O(N).
@author   Team 5 김주경, 서상원
@date     2025-11-22
"""

from typing import List
import random


class Solution:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        # Index of the k-th largest element if the array were sorted ascending
        target = len(nums) - k
        left, right = 0, len(nums) - 1

        while True:
            # If only one element is left, that must be the answer
            if left == right:
                return nums[left]

            # 1) Choose a random pivot index and value within the current range
            pivot_index = random.randint(left, right)
            pivot = nums[pivot_index]

            # 2) Dutch National Flag 3-way partition around the pivot
            #    After partitioning:
            #      nums[left .. lt-1]  : elements < pivot
            #      nums[lt   .. gt]    : elements == pivot
            #      nums[gt+1 .. right] : elements > pivot
            lt, i, gt = left, left, right
            while i <= gt:
                if nums[i] < pivot:
                    # Move smaller elements to the left region
                    nums[lt], nums[i] = nums[i], nums[lt]
                    lt += 1
                    i += 1
                elif nums[i] > pivot:
                    # Move larger elements to the right region
                    nums[gt], nums[i] = nums[i], nums[gt]
                    gt -= 1
                    # Do not increment i here, because the swapped-in value
                    # at position i has not been examined yet
                else:  # nums[i] == pivot
                    # Elements equal to pivot stay in the middle region
                    i += 1

            # 3) Decide which region contains the target index and shrink range
            if target < lt:
                # The target index lies in the "< pivot" region
                right = lt - 1
            elif target > gt:
                # The target index lies in the "> pivot" region
                left = gt + 1
            else:
                # The target index lies inside the "== pivot" region,
                # so nums[target] has the correct rank (k-th largest)
                return nums[target]
